File: JAVA/level2/구명보트.java
import java.util.*;

class Solution {
    public int solution(int[] people, int limit) {
        int answer = 0;
        Deque<Integer> cache = new ArrayDeque<Integer>();

        Arrays.sort(people);

        for(int i : people)
            cache.add(i);

        while(!cache.isEmpty()){
            if(cache.size() == 1 || cache.peekFirst() + cache.peekLast() > limit)
                cache.pollLast();
            else{
                int sum = limit - cache.pollFirst() - cache.pollLast();

                while(!cache.isEmpty() && sum - cache.peekFirst() >= 0)
                    sum -= cache.pollFirst();
            }
            answer++;
        }

        return answer;
    }
}

File: JavaScript/level2/구명보트.js
//효율성 시간초과
function numberSort(a, b){
    return b-a;
}

function solution(people, limit) {
    var answer = 0;
    
    people.sort(numberSort);
    
    while(people.length !== 0){
        let val = limit - people.shift();
        
        while(people[people.length-1] <= val){
            val -= people.pop();
        }
        answer++;
    }
    
    return answer;
}